### PR TITLE
Authorization rules

### PR DESCRIPTION
**Pull request by rashedarman**
**Implemented Changes**
- Install CanCanCan in your project.
- Add a `role` column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column `role` has value `"admin"`).
- A user can delete a comment if it is theirs or if they have an admin role (column `role` has value `"admin"`).

### CODE REVIEW

**Review by prolajumokeoni — COMMENTED** [review]
Hi @rashedarman 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request
You have submitted a pair-programming project, that does not have commit from your partner 👍🏿

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -18,6 +18,13 @@ def create
...
+  def destroy
+    @comment = Comment.find(params[:id]).destroy
+    respond_to do |f|
+      f.html { redirect_to user_posts_path(current_user), notice: 'The Comment has been destroyed successfully' }
+    end
+  end
+
   def comment_params
     params.require(:comment).permit(:text)
   end
```
Comment (Whoistolu) [inline app/controllers/comments_controller.rb:30]: - [ ] [REQUIRED]  The controller as it is currently written is primarily designed to handle HTML requests and responses. The `CommentsController` is not a standard endpoint that explicitly receives JSON input or responds with JSON output at the moment.
In the create action, when @comment.save is successful, it performs a redirect to the user_post_path with an HTML format response. And when there are errors during comment creation, it responds with JSON containing the error messages and the status 422 Unprocessable Entity. However, there is no mechanism in this controller to explicitly handle JSON input for comment creation.

To create a standard JSON API endpoint, you would typically have a separate set of actions that are dedicated to handling JSON requests and responses. For example, you might have create and destroy actions that are specifically designed to handle JSON input and return JSON output.

To make this controller a standard JSON API endpoint, you would need to modify it to handle JSON input (using data from request bodies) and respond with JSON output for each of the actions. 

### Let me modify your code to respond in json so you can use that as an example for other controllers

``` ruby
class CommentsController < ApplicationController
  # ... (other actions)

  def create
    @user = User.find(params[:user_id])
    @post = @user.posts.find(params[:post_id])
    @comment = Comment.new(comment_params)
    @comment.author = current_user
    @comment.post = @post

    if @comment.save
      redirect_to user_post_path(@user, @post), notice: 'Comment created successfully'
    else
      render json: { errors: @comment.errors.full_messages }, status: :unprocessable_entity
    end
  end

  def destroy
    @comment = Comment.find(params[:id]).destroy
    render json: { message: 'Comment destroyed successfully' }, status: :ok
  end

  # ... (comment_params and other methods)
end
```

### Let's modify other actions in other controllers as all the actions are not currently in the file changes.

**Review by prolajumokeoni — APPROVED** [review]
Hi @rashedarman and @Danfall369 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
- App has CanCanCan `ability` class that defines the set of rules for user  and post interactions  ✅ 
 - `user.admin?` method created inside the  model 🚀  

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -34,6 +36,13 @@ def create
     end
   end
 
+  def destroy
+    @post = Post.find(params[:id]).destroy
+    respond_to do |f|
+      f.html { redirect_to user_posts_path(current_user), notice: 'The Post has been destroyed successfully' }
+    end
+  end
```
Comment (Whoistolu) [inline app/controllers/posts_controller.rb:44]: - [ ] [REQUIRED] The destroy action in the provided code snippet is not designed as an API endpoint that receives JSON input or responds with JSON output. The action is specifically handling HTML requests and responding with an HTML redirect.

To create an API endpoint that handles JSON requests and responds with JSON output, you would need to modify the code to accommodate JSON input and output. Here's an example of how you could adapt the destroy action to work as a JSON API endpoint:

``` ruby
class PostsController < ApplicationController
  # ... (other actions)

  def destroy
    @post = Post.find_by(id: params[:id])
    
    if @post
      @post.destroy
      render json: { message: 'The Post has been destroyed successfully' }, status: :ok
    else
      render json: { error: 'Post not found' }, status: :not_found
    end
  end

  # ... (other actions)
end
```
### Let's modify other actions in other controllers as all the actions are not currently in the file changes.

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi @rashedarman and @Danfall369 
Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ]  _Check the inline comments under the review for sections to improve._

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**